File: dtb-api/store-service/src/main/java/app/roaim/dtbazar/storeservice/domain/Store.java
package app.roaim.dtbazar.storeservice.domain;

import app.roaim.dtbazar.storeservice.model.Image;
import lombok.Data;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.PersistenceConstructor;
import org.springframework.data.mongodb.core.index.GeoSpatialIndexed;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;
import java.util.Arrays;

@Data
@Document(collection = "store")
public class Store {

    @Id
    private String id;
    private String name;
    private String proprietor;
    private String mobile;
    private String address;
    private Image storeFrontImage;
    private String uid;
    @GeoSpatialIndexed
    private double[] location;
    private double totalDonation;
    private double spentDonation;
    private double allFoodPrice;
    @Indexed
    private boolean enabled = true;
    @CreatedDate
    private LocalDateTime createdAt = LocalDateTime.now();

    public synchronized void onIncreaseAllFoodPrice(double amount) {
        setAllFoodPrice(getAllFoodPrice() + amount);
    }

    public synchronized void onDecreaseAllFoodPrice(double amount) {
        setAllFoodPrice(getAllFoodPrice() - amount);
    }

    public synchronized void onSpentDonation(double amount) {
        setSpentDonation(getSpentDonation() + amount);
    }

    public synchronized void onAddedDonation(double amount) {
        setTotalDonation(getTotalDonation() + amount);
    }

    @PersistenceConstructor
    public Store(String name, String proprietor, String mobile, String address, Image storeFrontImage, String uid, double[] location) {
        super();
        update(name, proprietor, mobile, address, storeFrontImage, uid, location);
    }

    public Store(String name, String proprietor, String mobile, String address, Image storeFrontImage, String uid, double x, double y) {
        this(name, proprietor, mobile, address, storeFrontImage, uid, new double[]{x, y});
    }

    public void update(Store store) {
        update(store.getName(), store.getProprietor(), store.getMobile(), store.getAddress(), store.getStoreFrontImage(), store.getUid(), store.getLocation());
    }

    private void update(String name, String proprietor, String mobile, String address, Image storeFrontImage, String uid, double[] location) {
        this.name = name;
        this.proprietor = proprietor;
        this.mobile = mobile;
        this.address = address;
        this.storeFrontImage = storeFrontImage;
        this.uid = uid;
        this.location = location;
    }

    @Override
    public String toString() {
        return "Store [id= " + id + ", name= " + name + ", location= "
                + Arrays.toString(location) + "]";
    }
}

File: dtb-android/app/src/main/java/app/roaim/dtbazar/model/Store.kt
package app.roaim.dtbazar.model

import androidx.room.Embedded
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.Index
import com.google.gson.annotations.SerializedName

@Entity(tableName = "store", primaryKeys = ["id"], indices = [Index("uid")])
data class Store(

    @field:SerializedName("address")
    val address: String? = null,

    @field:SerializedName("uid")
    val uid: String,

    @field:SerializedName("createdAt")
    val createdAt: String? = null,

    @field:SerializedName("spentDonation")
    val spentDonation: Double? = null,

    @field:SerializedName("name")
    val name: String? = null,

    @field:SerializedName("mobile")
    val mobile: String? = null,

    @field:SerializedName("totalDonation")
    val totalDonation: Double? = null,

    @field:SerializedName("location")
    val location: List<Double?>? = null,

    @field:SerializedName("id")
    val id: String,

    @field:SerializedName("proprietor")
    val proprietor: String? = null,

    @Embedded(prefix = "sfi_")
    @field:SerializedName("storeFrontImage")
    val storeFrontImage: StoreFrontImage? = null,

    @field:SerializedName("allFoodPrice")
    val allFoodPrice: Double? = null
) : ListItem {
    @Ignore
    override fun getItemId(): String = id

    @Ignore
    fun totalDonationFloat() = totalDonation?.toFloat()

    @Ignore
    fun spentDonationFloat() = spentDonation?.toFloat()
}

data class StoreFrontImage(

    @field:SerializedName("id")
    val id: String? = null,

    @field:SerializedName("source")
    val source: String? = null,

    @field:SerializedName("url")
    val url: String? = null
)

data class StorePostBody(

    @field:SerializedName("address")
    val address: String? = null,

    @field:SerializedName("mobile")
    val mobile: String? = null,

    @field:SerializedName("name")
    val name: String? = null,

    @field:SerializedName("storeFrontImage")
    val storeFrontImage: StoreFrontImage? = null,

    @field:SerializedName("location")
    val location: List<Double?>? = null
)
